JBPM-6999 - Kie server client load balancer using wrong URL under concurrent access

MBJAVA opened the pull request:
I have added more details in here https://issues.jboss.org/browse/JBPM-6230

mswiderski commented:
ok to test

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+            public void run() {
+                try {
+                    startLatch.await();
+                    System.out.println("Th#" + threadNo + " Calling Kie Server ");
```
could you please replace system out with logger?

MBJAVA replied:
logger is updated

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+                    }
+
+                } catch (Exception e) {
+                    e.printStackTrace();
```
use logger to print the stacktrace

MBJAVA replied:
added logger

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+                } catch (Exception e) {
+                    e.printStackTrace();
+                } finally {
+                    System.out.println("Th#" + threadNo +" Done.");
```
use logger

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+
+        // Kickoff first scenario
+        int threadCount=2;
+        System.out.println("Starting 2 Threads");
```
use logger

MBJAVA replied:
added logger

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+
+        // We expect the threads to complete within 7 seconds
+        stopLatch.await(7, TimeUnit.SECONDS);
+        System.out.println("\nEnd of Threads - ");
```
use logger

MBJAVA replied:
added logger

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+        System.out.println("\nEnd of Threads - ");
+
+        // Need to sleep to ensure background thread completes.
+        System.out.println("\nSleeping for 3 seconds - ");
```
use logger

MBJAVA replied:
added logger

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+                                "  </kie-server-info>\n" +
+                                "</response>"))
+                .willSetStateTo("After success 2"));
+        System.out.println(" Current wireMockServer1 stub count =" + wireMockServer1.listAllStubMappings().getMappings().size());
```
use logger

MBJAVA replied:
added logger

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+        availableList.forEach(item -> System.out.println("Available Endpoint : [" + item + "]"));
+        failedList = ((AbstractKieServicesClientImpl)client).getLoadBalancer().getFailedEndpoints();
+        if (failedList.isEmpty()) {
+            System.out.println("Failed Endpoint : []");
```
use logger

MBJAVA replied:
added logger

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-client/src/test/java/org/kie/server/client/LoadBalancerClientTest.java:
```diff
@@ -290,6 +293,232 @@ public void testDefaultLoadBalancerNotValidHost() throws Exception {
+            System.out.println("Failed Endpoint : []");
+        }
+        else {
+            failedList.forEach(item -> System.out.println("Failed Endpoint : [" + item + "]"));
```
use logger

MBJAVA replied:
added logger

MBJAVA commented:
I have added logger and updated system.out.println with logger. If we don't use thread.sleep then only way to reproduce this issue is create heavy load of concurrent requests and failure. I have used Thread.sleep to replicate scenario in which second failure is updating failed url before url comes online.

mswiderski commented:
@MBJAVA just apply changes requested by @sutaakar and it will be ready to merge, thanks

MBJAVA commented:
I have configured smaller timeout and reduced fixedDelay. I also want to merge this change to 7.4.x branch. should I send another pull request for that? please let me know.

mswiderski commented:
Jenkins retest this

mswiderski commented:
Jenkins retest this

mswiderski commented:
@MBJAVA thanks for the changes, we just need to wait for the build to be green (don't worry if you will see there are 4 failures right now as there are some tests broken in master - once they are fixed I'll merge this one). 
When it comes to 7.4.x we don't maintain already released community branches so you would have to patch 7.4 yourself locally only or upgrade to 7.7 that will include this fix.

mswiderski commented:
Jenkins retest this

tsurdilo commented:
Jenkins retest please

mswiderski commented:
Jenkins retest this

mswiderski commented:
Jenkins retest this

mswiderski commented:
@MBJAVA there is a test failure that is certainly related:

```
org.kie.server.common.rest.NoEndpointFoundException: No available endpoints found
	at org.kie.server.client.balancer.impl.AbstractBalancerStrategy.checkEmpty(AbstractBalancerStrategy.java:27)
	at org.kie.server.client.balancer.impl.RoundRobinBalancerStrategy.next(RoundRobinBalancerStrategy.java:37)
	at org.kie.server.client.balancer.LoadBalancer.getUrl(LoadBalancer.java:52)
	at org.kie.server.client.impl.AbstractKieServicesClientImpl.invoke(AbstractKieServicesClientImpl.java:809)
	at org.kie.server.client.impl.AbstractKieServicesClientImpl.makeHttpGetRequestAndCreateServiceResponse(AbstractKieServicesClientImpl.java:177)
	at org.kie.server.client.impl.KieServicesClientImpl.getServerState(KieServicesClientImpl.java:280)
	at org.kie.server.client.LoadBalancerClientTest.testMultipleConcurrentFailRequestsForLoadBalancerWithSingleServer(LoadBalancerClientTest.java:349)
```

could you please take a look?

MBJAVA commented:
I don't get this exception so I am not able to test but I have increased time-out little bit if that can resolve the issue.

mswiderski commented:
Jenkins retest this

MBJAVA commented:
is jenkin build still failing? should we go back to existing time out value?

mswiderski commented:
@MBJAVA let's give it a try with the original values

MBJAVA commented:
what is the failure here? I don't get that same exception in my workspace when I build this project. I can put back original values but that does make test slow. Trying to figure out if there is any other way to solve this issue.

mswiderski commented:
Jenkins retest this

mswiderski commented:
@MBJAVA please do, we run it couple of times to see if that's stable

MBJAVA commented:
This bug was found in 7.4.1.Final, can you add this fix and create minor version that we can use.

MBJAVA commented:
When are you planning to merge this change?

mswiderski commented:
it's good from my point of view, @sutaakar wdyt?

sutaakar commented:
@mswiderski good for me.

mswiderski commented:
@MBJAVA thanks for your contribution and patience :)
